cmd/symbols/internal/parser/parser_pool.go

```diff
@@ -51,7 +83,57 @@ func NewParserPool(newParser ParserFactory, numParserProcesses int, parserTypes
 // the pool. This method always returns a non-nil parser with a nil error value.
 //
 // This method blocks until a parser is available or the given context is canceled.
-func (p *parserPool) Get(ctx context.Context, source ctags_config.ParserType) (ctags.Parser, error) {
+func (p *ParserPool) GetParser(ctx context.Context, path string, content []byte) (ctags.Parser, ctags_config.ParserType, error) {
+
+	parserType := p.getParserType(path, content)
+	if ctags_config.ParserIsNoop(parserType) {
+		return nil, parserType, nil
+	}
+
+	parser, err := p.parserFromPool(ctx, parserType)
+	if err != nil {
+		return nil, parserType, err
+	}
+
+	return parser, parserType, nil
+}
+
+func (p *ParserPool) Done(parser ctags.Parser, source ctags_config.ParserType) {
+	pool := p.pool[source]
+	pool <- parser
+}
+
+func (p *ParserPool) parserFromPool(ctx context.Context, source ctags_config.ParserType) (ctags.Parser, error) {
+	if ctags_config.ParserIsNoop(source) {
+		return nil, errors.New("Should not pass Noop ParserType to this function")
+	}
+
+	p.metrics.parseQueueSize.Inc()
+	defer p.metrics.parseQueueSize.Dec()
+
+	parser, err := p.get(ctx, source)
+	if err != nil {
+		if ctx.Err() == err {
+			p.metrics.parseQueueTimeouts.Inc()
+		} else {
+			err = errors.Wrap(err, "failed to create parser")
+		}
+	}
```

Comment:
also idiomatic go to do early return

```suggestion
	if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			p.metrics.parseQueueTimeouts.Inc()
		}
        // ignore err if context has expired since err is likely due to that
        if ctx.Err() != nil {
            return nil, ctx.Err()
        }
        return nil, errors.Wrap(err, "failed to create parser")
    }        
```